src/units/abilities.cpp

```diff
@@ -1168,52 +1168,120 @@ void attack_type::modified_attacks(unsigned & min_attacks,
+				max_res = std::max(max_res, type_res[type]);
 			}
-			std::sort( std::begin( type_list ) , std::end( type_list ) , [&]( const std::string& rhs , const std::string& lhs ){
-				return type_count[lhs] < type_count[rhs];
-			});
 		}
 	}
-	if(!type_list.empty()){
-		return type_list.front();
+
+	if (type_res.empty()) return "";
+
+	std::vector<std::string> type_list;
+	for(auto& i : type_res){
+		if(i.second == max_res){
+			type_list.push_back(i.first);
+		}
 	}
-	return std::nullopt;
+	if(!type_list.empty()) return "";
+
+	if(type_list.size() == 1) return type_list.front();
+	std::sort(type_list.begin(), type_list.end());
+
+	return type_list.front();
+}
+
+std::string attack_type::select_damage_type(const unit_ability_list& damage_type_list, const std::string& key_name, unit_ability_list resistance_list) const
+{
+	bool is_alternative = (key_name == "alternative_type");
+	if(is_alternative && other_){
+		return select_alternative_type(damage_type_list, resistance_list, (*other_));
+	} else if(!is_alternative){
+		return select_replacement_type(damage_type_list);
+	}
+	return "";
 }
 
 /**
  * Returns the type of damage inflicted.
  */
 std::pair<std::string, std::string> attack_type::damage_type() const
 {
-	unit_ability_list abil_list = get_specials_and_abilities("damage_type");
-	if(abil_list.empty()){
+	unit_ability_list damage_type_list = get_specials_and_abilities("damage_type");
+	if(damage_type_list.empty()){
 		return {type(), ""};
 	}
 
-	std::optional<std::string> replacement_type = select_damage_type(abil_list, "replacement_type");
-	std::optional<std::string> alternative_type = select_damage_type(abil_list, "alternative_type");
-	std::string type_damage = replacement_type.value_or(type());
-	if(alternative_type && type_damage != *alternative_type){
-		return {type_damage, *alternative_type};
+	unit_ability_list resistance_list;
+	if(other_){
+		resistance_list = (*other_).get_abilities_weapons("resistance", other_loc_, other_attack_, shared_from_this());
+	}
+	std::string replacement_type = select_damage_type(damage_type_list, "replacement_type", resistance_list);
+	std::string alternative_type = select_damage_type(damage_type_list, "alternative_type", resistance_list);
+	std::string type_damage = replacement_type.empty() ? type() : replacement_type;
+	if(!alternative_type.empty() && type_damage != alternative_type){
+		return {type_damage, alternative_type};
 	}
 	return {type_damage, ""};
 }
 
+std::set<std::string> attack_type::damage_alternative_type() const
+{
+	unit_ability_list damage_alternative_type_list = get_specials_and_abilities("damage_type");
+	if(damage_alternative_type_list.empty()){
+		return {};
+	}
+	std::set<std::string> type_count;
+	for(auto& i : damage_alternative_type_list) {
+		const config& c = *i.ability_cfg;
+		if(!c["alternative_type"].empty() && type_count.count(c["alternative_type"].str()) == 0) {
+			type_count.insert(c["alternative_type"].str());
+		}
```

Comment:
it can't added two identical element in a set?

Reply:
and if i don't check if "alternative_type" is empty i have a show bug, like for resistance_value who evaluate un type who don't exixt because empty don't checked